vllm/v1/attention/backends/mps_attn.py

```diff
@@ -0,0 +1,615 @@
+        key: torch.Tensor,
+        value: torch.Tensor,
+        kv_cache: torch.Tensor,
+        attn_metadata: MPSAttentionMetadata | None,
+        output: torch.Tensor | None = None,
+        output_scale: torch.Tensor | None = None,
+        output_block_scale: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        """Forward pass for MPS attention backend.
+
+        Args:
+            query: shape = [num_tokens, num_heads, head_size]
+            key: shape = [num_tokens, num_kv_heads, head_size]
+            value: shape = [num_tokens, num_kv_heads, head_size]
+            kv_cache: shape = [2, num_blocks, num_kv_heads, block_size, head_size]
+            attn_metadata: Metadata for attention.
+
+        Returns:
+            output tensor with shape = [num_tokens, num_heads, head_size]
+        """
+        assert output is not None, "Output tensor must be provided."
+        if output_scale is not None or output_block_scale is not None:
+            raise NotImplementedError(
+                "Fused output quantization is not yet supported"
+                " for MPSAttentionBackendImpl"
+            )
+
+        # For warming-up / profiling
+        if attn_metadata is None:
+            return output
+
+        num_actual_tokens = attn_metadata.num_actual_tokens
+
+        # Handle encoder attention - no KV cache
+        if self.attn_type in (AttentionType.ENCODER_ONLY, AttentionType.ENCODER):
+            return self._compute_attention_no_cache(
+                query[:num_actual_tokens],
+                key[:num_actual_tokens],
+                value[:num_actual_tokens],
+                output[:num_actual_tokens],
+                attn_metadata,
+                is_causal=False,
+            )
+
+        # For decoder attention, always use KV cache
+        key_cache, value_cache = kv_cache.unbind(0)
+
+        # Update KV cache with new key/value tensors using vectorized ops
+        if (
+            self.kv_sharing_target_layer_name is None
+            and key is not None
+            and value is not None
+        ):
+            self._update_kv_cache_vectorized(
+                key[:num_actual_tokens],
+                value[:num_actual_tokens],
+                key_cache,
+                value_cache,
+                attn_metadata.slot_mapping[:num_actual_tokens],
+            )
+
+        # Compute attention
+        # For prefill (max_query_len == max_seq_len): use original Q/K/V directly
+        # For decode (max_query_len == 1): read K/V from cache
+        max_query_len = attn_metadata.max_query_len
+        max_seq_len = attn_metadata.max_seq_len
+
+        if max_query_len == max_seq_len:
+            # Prefill: use original tensors directly for accuracy
+            return self._compute_attention_no_cache(
+                query[:num_actual_tokens],
+                key[:num_actual_tokens],
+                value[:num_actual_tokens],
+                output[:num_actual_tokens],
```

Comment:
The CPU version begins this slice at `num_decode_tokens` in prefill, why don't we do that here?

Reply:
Changed prefill to slice from num_decode_tokens to num_actual_tokens instead of 0 to num_actual_tokens